ramips: add support for MTS WG430223

csharper2005 opened the pull request:
MTS WG430223 is a wireless AC1300 (WiFi 5) router manufactured by Arcadyan company. It's very similar to Beeline Smartbox Flash (Arcadyan WG443223).

Device specification
--------------------
SoC Type: MediaTek MT7621AT
RAM: 128 MiB
Flash: 128 MiB (Winbond W29N01HV)
Wireless 2.4 GHz (MT7615DN): b/g/n, 2x2
Wireless 5 GHz (MT7615DN): a/n/ac, 2x2
Ethernet: 3xGbE (WAN, LAN1, LAN2)
USB ports: No
Button: 1 (Reset/WPS)
LEDs: 2 (Red, Green)
Power: 12 VDC, 1 A
Connector type: Barrel
Bootloader: U-Boot (Ralink UBoot Version: 5.0.0.2)
OEM: Arcadyan WG430223

Installation
------------
1. Login to the router web interface (superadmin:serial number)
2. Navigate to Administration -> Miscellaneous -> Access control lists & enable telnet & enable "Remote control from any IP address"
3. Connect to the router using telnet (default admin:admin)
4. Place *factory.trx on any web server (192.168.1.2 in this example)
5. Connect to the router using telnet shell (no password required)
6. Save MAC adresses to U-Boot environment:
```
   uboot_env --set --name eth2macaddr --value $(ifconfig | grep eth2 | \
    awk '{print $5}')
   uboot_env --set --name eth3macaddr --value $(ifconfig | grep eth3 | \
    awk '{print $5}')
   uboot_env --set --name ra0macaddr --value $(ifconfig | grep ra0 | \
    awk '{print $5}')
   uboot_env --set --name rax0macaddr --value $(ifconfig | grep rax0 | \
    awk '{print $5}')
```
7. Ensure that MACs were saved correctly:
```   
   uboot_env --get --name eth2macaddr
   uboot_env --get --name eth3macaddr
   uboot_env --get --name ra0macaddr
   uboot_env --get --name rax0macaddr
```
8. Download and write the OpenWrt images:
```
   cd /tmp
   wget http://192.168.1.2/factory.trx
   mtd_write erase /dev/mtd4
   mtd_write write factory.trx /dev/mtd4
```
10. Set 1st boot partition and reboot:
```
   uboot_env --set --name bootpartition --value 0
```
Back to Stock
-------------
1. Run in the OpenWrt shell:
```
   fw_setenv bootpartition 1
   reboot
```
2. Optional step. Upgrade the stock firmware with any version to overwrite the OpenWrt in Slot 1.

MAC addresses
-------------
```
+-----------+-------------------+----------------+
| Interface | MAC               | Source         |
+-----------+-------------------+----------------+
| label     | A4:xx:xx:51:xx:F4 | No MACs was    |
| LAN       | A4:xx:xx:51:xx:F6 | found on Flash |
| WAN       | A4:xx:xx:51:xx:F4 | [1]            |
| WLAN_2g   | A4:xx:xx:51:xx:F5 |                |
| WLAN_5g   | A6:xx:xx:21:xx:F5 |                |
+-----------+-------------------+----------------+
```
[1]:
a. Label wasb't found neither in factory nor in other places.
b. MAC addresses are stored in encrypted partition "glbcfg". Encryption key hasn't known yet. To ensure the correct MACs in OpenWrt, a hack with saving of the MACs to u-boot-env during the installation was applied.
c. Default Ralink ethernet MAC address (00:0C:43:28:80:A0) was found in "Factory" 0xfff0. It's the same for all MTS WG430223 devices. OEM firmware also uses this MAC when initialazes ethernet driver. In OpenWrt we use it only as internal GMAC (eth0), all other MACs are unique. Therefore, there is no any barriers to the operation of several MTS WG430223 devices even within the same broadcast domain.

Stock firmware image format
---------------------------
The same as Beeline Smartbox Flash but with another trx magic
```
+--------------+---------------+----------------------------------------+
| Offset       |               | Description                            |
+==============+===============+========================================+
| 0x0          | 31 52 48 53   | TRX magic "1RHS"                       |
+--------------+---------------+----------------------------------------+
```

csharper2005 committed:
ramips: allow custom trx magic for Arcadyan

This commit:
1. Renames beeline-trx recipe in mt7621.mk to arcadyan-trx. The recipe
   is necessary for:
   - MTS WG430223 (Arcadyan WG430223)
   - Beeline Smartbox Flash (Arcadyan WG443223)
2. Allows specify custom trx magic which is different for the routers
   mentined above.

Signed-off-by: Mikhail Zhilkin <csharper2005@gmail.com>

981213 commented on target/linux/ramips/dts/mt7621_mts_wg430223.dts:
```diff
@@ -0,0 +1,195 @@
+&gmac0 {
+	nvmem-cells = <&macaddr_factory_fff0>;
+	nvmem-cell-names = "mac-address";
+};
```
If there's no valid mac here, this part can be removed.

csharper2005 replied:
> If there's no valid mac here, this part can be removed.

Yeah, we can safely drop this part. We discussed this point with @mans0n in Beeline SmartBox flash PR. We have default ralink Mac here. In favor of keeping:
1. Default ralink Mac would be better than the random one, which changes on every boot;
2. End user can easily check that the eeprom partition is "alive".
What do you think about this? I'm ready to drop this code if you still think it's redundant.

981213 replied:
I'm fine either way. If you keep it, maybe add some comments in dts saying there's only the ralink default mac here?

981213 reviewed (commented):
Hi! You can reorder the last two commits to avoid adding a bunch of lines and immediately removing them.

One more comment besides that:

csharper2005 commented:
> Hi! You can reorder the last two commits to avoid adding a bunch of lines and immediately removing them.

Hi, @981213! Thanks for the review! Got it.

csharper2005 committed:
ramips: create new dtsi for Arcadyan routers

This commit moves common properties for the boards below to a new dtsi:
Beeline Smartbox Flash (Arcadyan WG443223)
MTS WG430223 (Arcadyan WG430223)

The boards are almost the same. Here is the differences:
+------+----------+----------+
|      | WG430223 | WG443223 |
+------+----------+----------+
| RAM  | 128      | 256      |
+------+----------+----------+
| USB  | -        | 1x3.0    |
+------+----------+----------+
| LEDS | RG       | RGB      |
+------+----------+----------+

Signed-off-by: Mikhail Zhilkin <csharper2005@gmail.com>

csharper2005 committed:
ramips: add support for MTS WG430223

MTS WG430223 is a wireless AC1300 (WiFi 5) router manufactured by
Arcadyan company. It's very similar to Beeline Smartbox Flash (Arcadyan
WG443223).

Device specification
--------------------
SoC Type: MediaTek MT7621AT
RAM: 128 MiB
Flash: 128 MiB (Winbond W29N01HV)
Wireless 2.4 GHz (MT7615DN): b/g/n, 2x2
Wireless 5 GHz (MT7615DN): a/n/ac, 2x2
Ethernet: 3xGbE (WAN, LAN1, LAN2)
USB ports: No
Button: 1 (Reset/WPS)
LEDs: 2 (Red, Green)
Power: 12 VDC, 1 A
Connector type: Barrel
Bootloader: U-Boot (Ralink UBoot Version: 5.0.0.2)
OEM: Arcadyan WG430223

Installation
------------
1. Login to the router web interface (superadmin:serial number)
2. Navigate to Administration -> Miscellaneous -> Access control lists &
   enable telnet & enable "Remote control from any IP address"
3. Connect to the router using telnet (default admin:admin)
4. Place *factory.trx on any web server (192.168.1.2 in this example)
5. Connect to the router using telnet shell (no password required)
6. Save MAC adresses to U-Boot environment:
   uboot_env --set --name eth2macaddr --value $(ifconfig | grep eth2 | \
    awk '{print $5}')
   uboot_env --set --name eth3macaddr --value $(ifconfig | grep eth3 | \
    awk '{print $5}')
   uboot_env --set --name ra0macaddr --value $(ifconfig | grep ra0 | \
    awk '{print $5}')
   uboot_env --set --name rax0macaddr --value $(ifconfig | grep rax0 | \
    awk '{print $5}')
7. Ensure that MACs were saved correctly:
   uboot_env --get --name eth2macaddr
   uboot_env --get --name eth3macaddr
   uboot_env --get --name ra0macaddr
   uboot_env --get --name rax0macaddr
8. Download and write the OpenWrt images:
   cd /tmp
   wget http://192.168.1.2/factory.trx
   mtd_write erase /dev/mtd4
   mtd_write write factory.trx /dev/mtd4
9. Set 1st boot partition and reboot:
   uboot_env --set --name bootpartition --value 0

Back to Stock
-------------
1. Run in the OpenWrt shell:
   fw_setenv bootpartition 1
   reboot
2. Optional step. Upgrade the stock firmware with any version to
   overwrite the OpenWrt in Slot 1.

MAC addresses
-------------
+-----------+-------------------+----------------+
| Interface | MAC               | Source         |
+-----------+-------------------+----------------+
| label     | A4:xx:xx:51:xx:F4 | No MACs was    |
| LAN       | A4:xx:xx:51:xx:F6 | found on Flash |
| WAN       | A4:xx:xx:51:xx:F4 | [1]            |
| WLAN_2g   | A4:xx:xx:51:xx:F5 |                |
| WLAN_5g   | A6:xx:xx:21:xx:F5 |                |
+-----------+-------------------+----------------+
[1]:
a. Label wasb't found neither in factory nor in other places.
b. MAC addresses are stored in encrypted partition "glbcfg". Encryption
   key hasn't known yet. To ensure the correct MACs in OpenWrt, a hack
   with saving of the MACs to u-boot-env during the installation was
   applied.
c. Default Ralink ethernet MAC address (00:0C:43:28:80:A0) was found in
   "Factory" 0xfff0. It's the same for all MTS WG430223 devices. OEM
   firmware also uses this MAC when initialazes ethernet driver. In
   OpenWrt we use it only as internal GMAC (eth0), all other MACs are
   unique. Therefore, there is no any barriers to the operation of several
   MTS WG430223 devices even within the same broadcast domain.

Stock firmware image format
---------------------------
The same as Beeline Smartbox Flash but with another trx magic
+--------------+---------------+----------------------------------------+
| Offset       |               | Description                            |
+==============+===============+========================================+
| 0x0          | 31 52 48 53   | TRX magic "1RHS"                       |
+--------------+---------------+----------------------------------------+

Signed-off-by: Mikhail Zhilkin <csharper2005@gmail.com>

csharper2005 commented:
Hi, @981213! 

Commits have been reordered according to your recommendation. 

Please, let me know if your opinion regarding default mac has changed (I haven't removed that code block yet).

981213 commented:
Merged into my staging tree at: https://git.openwrt.org/?p=openwrt/staging/981213.git
Thanks!

csharper2005 commented:
@981213 Thank you!

musashino205 commented:
@csharper2005 This is just a question,
Buffalo sells/sold some routers manufactured by Arcadyan and they have partition structure close to your device. And, device-specific MAC address is stored in ```board_data``` partition with the key ```mac=```.

example (WSR-2533DHP2):

```
$ hexdump -n 128 -C mtd11_board_data.bin
00000000  6d 61 63 3d 31 38 3a 43  32 3a 42 46 3a 38 45 3a  |mac=18:C2:BF:8E:|
00000010  46 35 3a 38 38 0a 63 63  3d 4a 50 0a 68 77 72 65  |F5:88.cc=JP.hwre|
00000020  76 3d 30 0a 70 69 6e 3d  XX XX XX XX XX XX XX XX  |v=0.pin=XXXXXXXX|
00000030  0a 6b 65 79 3d XX XX XX  XX XX XX XX XX XX XX XX  |.key=XXXXXXXXXXX|
00000040  XX XX 0a 70 61 73 73 77  6f 72 64 3d XX XX XX XX  |XX.password=XXXX|
00000050  XX XX XX XX 0a 00 00 00  00 00 00 00 00 00 00 00  |XXXX............|
00000060  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000080
```

Does your device has no MAC address in the partition?

csharper2005 commented:
@musashino205 unfortunately no. These arcadyans are more advanced (lol) and keep MAC addresses and passwords in encrypted partitions "glbcfg" or / and "glbcfg2". I can share my mtd backups if you are interested in this topic.

musashino205 commented:
@csharper2005 Ah, very sad... 😭  Okay, thanks!

csharper2005 commented:
@musashino205 good news! Guys (@abyrga, @r3d5ky in telegram) found a way how to decrypt a partition [here](https://t.me/smartboxgiga/49944). For MTS WG430223 and MTS WG430223.

```
$ dd status=none if=mtd9 bs=1 count=368 skip=256 | openssl aes-128-cbc -d -nopad -K 2A4B303D7644395C3B2B7053553C5200 -iv 00000000000000000000000000000000
mac=ec:6c:##:##:##:b8
sn=AG##########
ssid=Beeline_2G_######
psk=###########
ssid2=Beeline_5G_######
psk2=###########
guest_ssid=Beeline_2G_######-Guest
guest_psk=###########
guest_ssid2=Beeline_5G_######-Guest
guest_psk2=###########
group_ssid=BH-Beeline_2G_######
group_psk=###########
group_ssid2=BH-Beeline_5G_######
group_psk2=###########
hw=01
password=#####
```

Ansuel commented:
@csharper2005 interesting thing... worth opening an issue about the subject so we can track this better?
